File: OpenHarmony/developtools/hapsigner/hapsigntool_cpp/hap/verify/src/verify_hap.cpp
/*
 * Copyright (c) 2024-2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <climits>
#include <cstdlib>
#include <regex>
#include <unordered_map>
#include <vector>
#include <algorithm>

#include "securec.h"
#include "hap_signer_block_utils.h"
#include "signature_info.h"
#include "options.h"
#include "openssl/pem.h"
#include "pkcs7_data.h"
#include "hap_utils.h"
#include "string_utils.h"
#include "verify_code_signature.h"
#include "param_constants.h"
#include "file_utils.h"
#include "nlohmann/json.hpp"
#include "verify_hap.h"

using namespace nlohmann;
namespace OHOS {
namespace SignatureTools {
const int32_t VerifyHap::HEX_PRINT_LENGTH = 3;
const int32_t VerifyHap::DIGEST_BLOCK_LEN_OFFSET = 8;
const int32_t VerifyHap::DIGEST_ALGORITHM_OFFSET = 12;
const int32_t VerifyHap::DIGEST_LEN_OFFSET = 16;
const int32_t VerifyHap::DIGEST_OFFSET_IN_CONTENT = 20;
const std::string VerifyHap::HAP_APP_PATTERN = "[^]*.hap$";
const std::string VerifyHap::HQF_APP_PATTERN = "[^]*.hqf$";
const std::string VerifyHap::HSP_APP_PATTERN = "[^]*.hsp$";
const std::string VerifyHap::APP_APP_PATTERN = "[^]*.app$";
static constexpr int ZIP_HEAD_OF_SUBSIGNING_BLOCK_LENGTH = 12;

VerifyHap::VerifyHap() : isPrintCert(true)
{
}

VerifyHap::VerifyHap(bool printCert)
{
    isPrintCert = printCert;
}

void VerifyHap::setIsPrintCert(bool printCert)
{
    isPrintCert = printCert;
}

bool VerifyHap::HapOutPutPkcs7(PKCS7* p7, const std::string& outPutPath)
{
    std::string p7bContent = StringUtils::Pkcs7ToString(p7);
    if (p7bContent.empty()) {
        SIGNATURE_TOOLS_LOGE("p7b to string failed!\n");
        return false;
    }
    if (FileUtils::Write(p7bContent, outPutPath) < 0) {
        SIGNATURE_TOOLS_LOGE("p7b write to file falied!\n");
        return false;
    }
    return true;
}

bool VerifyHap::outputOptionalBlocks(const std::string& outputProfileFile, const std::string& outputProofFile,
                                     const std::string& outputPropertyFile,
                                     const std::vector<OptionalBlock>& optionBlocks)
{
    for (auto& optionBlock : optionBlocks) {
        if (optionBlock.optionalType == HapUtils::HAP_PROFILE_BLOCK_ID) {
            if (!writeOptionalBytesToFile(optionBlock, outputProfileFile)) {
                return false;
            }
        } else if (optionBlock.optionalType == HapUtils::HAP_PROPERTY_BLOCK_ID) {
            if (!writeOptionalBytesToFile(optionBlock, outputPropertyFile)) {
                return false;
            }
        } else if (optionBlock.optionalType == HapUtils::HAP_PROOF_OF_ROTATION_BLOCK_ID) {
            if (!writeOptionalBytesToFile(optionBlock, outputProofFile)) {
                return false;
            }
        } else {
            SIGNATURE_TOOLS_LOGE("Unsupported Block Id: %d", optionBlock.optionalType);
            return false;
        }
    }
    return true;
}
bool VerifyHap::writeOptionalBytesToFile(const OptionalBlock& optionalBlock, const std::string& path)
{
    if (path.empty()) {
        return true;
    }
    std::string optionBlockString(optionalBlock.optionalBlockValue.GetBufferPtr(),
                          optionalBlock.optionalBlockValue.GetCapacity());
    if (FileUtils::Write(optionBlockString, path) < 0) {
        PrintErrorNumberMsg("IO_ERROR", IO_ERROR, "write optional bytes to file:" + path + " falied!");
        return false;
    }
    return true;
}

bool VerifyHap::HapOutPutCertChain(std::vector<X509*>& certs, const std::string& outPutPath)
{
    if (isPrintCert) {
        if (!PrintCertChainToCmd(certs)) {
            SIGNATURE_TOOLS_LOGE("print cert chain to cmd failed\n");
            return false;
        }
    }
    VerifyHapOpensslUtils::GetOpensslErrorMessage();
    SIGNATURE_TOOLS_LOGD("outPutPath = %s", outPutPath.c_str());
    std::vector<std::string> certStr;
    for (auto& cert : certs) {
        certStr.emplace_back(StringUtils::SubjectToString(cert));
        certStr.emplace_back(StringUtils::x509CertToString(cert));
    }
    std::string outPutCertChainContent = std::accumulate(certStr.begin(), certStr.end(), std::string(),
        [](std::string sum, const std::string& certstr) { return sum + certstr; });
    if (FileUtils::Write(outPutCertChainContent, outPutPath) < 0) {
        SIGNATURE_TOOLS_LOGE("certChain write to file falied!\n");
        return false;
    }
    return true;
}

bool VerifyHap::PrintCertChainToCmd(std::vector<X509*>& certChain)
{
    BIO* outFd = BIO_new_fp(stdout, BIO_NOCLOSE);
    if (!outFd) {
        PrintErrorNumberMsg("IO_ERROR", IO_ERROR, "The stdout stream may have errors");
        return false;
    }
    uint64_t format = XN_FLAG_SEP_COMMA_PLUS; // Print according to RFC2253
    uint64_t content = X509_FLAG_NO_EXTENSIONS | X509_FLAG_NO_ATTRIBUTES | X509_FLAG_NO_HEADER | X509_FLAG_NO_SIGDUMP;
    int num = 0;
    for (auto& cert : certChain) {
        PrintMsg("+++++++++++++++++++++++++++++++++certificate #" + std::to_string(num) +
                 "+++++++++++++++++++++++++++++++++++++");
        if (!X509_print_ex(outFd, cert, format, content)) {
            VerifyHapOpensslUtils::GetOpensslErrorMessage();
            SIGNATURE_TOOLS_LOGE("print x509 cert to cmd failed");
            BIO_free(outFd);
            return false;
        }
        ++num;
    }
    BIO_free(outFd);
    return true;
}

int32_t VerifyHap::Verify(const std::string& filePath, Options* options)
{
    SIGNATURE_TOOLS_LOGD("Start Verify");
    std::string standardFilePath;
    if (!CheckFilePath(filePath, standardFilePath)) {
        SIGNATURE_TOOLS_LOGE("Check file path%s failed", filePath.c_str());
        return IO_ERROR;
    }
    RandomAccessFile hapFile;
    if (!hapFile.Init(standardFilePath)) {
        SIGNATURE_TOOLS_LOGE("%s init failed", standardFilePath.c_str());
        return ZIP_ERROR;
    }
    int32_t resultCode = Verify(hapFile, options, filePath);
    if (resultCode != RET_OK) {
        PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR, standardFilePath + " verify failed");
    }
    return resultCode;
}

bool VerifyHap::CheckFilePath(const std::string& filePath, std::string& standardFilePath)
{
    char path[PATH_MAX] = { 0x00 };
    if (filePath.size() > PATH_MAX || realpath(filePath.c_str(), path) == nullptr) {
        PrintErrorNumberMsg("IO_ERROR", IO_ERROR,
                            filePath + " does not exist or is over " + std::to_string(PATH_MAX) + " chars");
        return false;
    }
    standardFilePath = std::string(path);
    std::string standardFilePathTmp = std::string(path);
    std::transform(standardFilePathTmp.begin(), standardFilePathTmp.end(), standardFilePathTmp.begin(),
                   [](unsigned char c) { return std::tolower(c); });
    bool ret = (!std::regex_match(standardFilePathTmp, std::regex(HAP_APP_PATTERN)) &&
                !std::regex_match(standardFilePathTmp, std::regex(HSP_APP_PATTERN)) &&
                !std::regex_match(standardFilePathTmp, std::regex(APP_APP_PATTERN)) &&
                !std::regex_match(standardFilePathTmp, std::regex(HQF_APP_PATTERN)));
    if (ret) {
        PrintErrorNumberMsg("COMMAND_PARAM_ERROR", COMMAND_PARAM_ERROR,
                            "only support format is [hap, hqf, hsp, app]");
        return false;
    }
    return true;
}

int32_t VerifyHap::Verify(RandomAccessFile& hapFile, Options* options, const std::string& filePath)
{
    SignatureInfo hapSignInfo;
    if (!HapSignerBlockUtils::FindHapSignature(hapFile, hapSignInfo)) {
        return ZIP_ERROR;
    }

    if (CheckCodeSign(filePath, hapSignInfo.optionBlocks) == false) {
        SIGNATURE_TOOLS_LOGE("check coode sign failed\n");
        return VERIFY_ERROR;
    }

    Pkcs7Context pkcs7Context;
    if (!VerifyAppPkcs7(pkcs7Context, hapSignInfo.hapSignatureBlock)) {
        return PARSE_ERROR;
    }

    if (!GetDigestAndAlgorithm(pkcs7Context)) {
        SIGNATURE_TOOLS_LOGE("Get digest failed");
        return PARSE_ERROR;
    }

    STACK_OF(X509_CRL)* x509Crl = nullptr;
    if (!VerifyHapOpensslUtils::GetCrlStack(pkcs7Context.p7, x509Crl)) {
        SIGNATURE_TOOLS_LOGE("Get Crl stack failed");
        return PARSE_ERROR;
    }

    if (!VerifyCertOpensslUtils::VerifyCrl(pkcs7Context.certChain[0], x509Crl, pkcs7Context)) {
        SIGNATURE_TOOLS_LOGE("Verify Crl stack failed");
        return VERIFY_ERROR;
    }

    if (!HapSignerBlockUtils::VerifyHapIntegrity(pkcs7Context, hapFile, hapSignInfo)) {
        SIGNATURE_TOOLS_LOGE("Verify Integrity failed");
        return VERIFY_ERROR;
    }
    if (!HapOutPutCertChain(pkcs7Context.certChain[0],
        options->GetString(Options::OUT_CERT_CHAIN))) {
        SIGNATURE_TOOLS_LOGE("out put cert chain failed");
        return IO_ERROR;
    }

    if (!outputOptionalBlocks(options->GetString(ParamConstants::PARAM_VERIFY_PROFILE_FILE),
                              options->GetString(ParamConstants::PARAM_VERIFY_PROOF_FILE),
                              options->GetString(ParamConstants::PARAM_VERIFY_PROPERTY_FILE),
                              hapSignInfo.optionBlocks)) {
        SIGNATURE_TOOLS_LOGE("output Optional Blocks failed");
        return IO_ERROR;
    }
    return RET_OK;
}

bool VerifyHap::CheckCodeSign(const std::string& hapFilePath,
                              const std::vector<OptionalBlock>& optionalBlocks)const
{
    std::unordered_map<int, ByteBuffer> map;
    for (const OptionalBlock& block : optionalBlocks) {
        map.emplace(block.optionalType, block.optionalBlockValue);
    }
    bool codeSignFlag = map.find(HapUtils::HAP_PROPERTY_BLOCK_ID) != map.end() &&
        map[HapUtils::HAP_PROPERTY_BLOCK_ID].GetCapacity() > 0;
    if (codeSignFlag) {
        ByteBuffer propertyBlockArray = map[HapUtils::HAP_PROPERTY_BLOCK_ID];
        std::vector<std::string> fileNameArray = StringUtils::SplitString(hapFilePath, '.');
        if (fileNameArray.size() < ParamConstants::FILE_NAME_MIN_LENGTH) {
            PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR, "ZIP64 format not supported.");
            return false;
        }

        if (propertyBlockArray.GetCapacity() < ZIP_HEAD_OF_SUBSIGNING_BLOCK_LENGTH)
            return false;
        uint32_t blockType;
        propertyBlockArray.GetUInt32(OFFSET_ZERO, blockType);
        uint32_t blockLength;
        propertyBlockArray.GetUInt32(OFFSET_FOUR, blockLength);
        uint32_t blockOffset;
        propertyBlockArray.GetUInt32(OFFSET_EIGHT, blockOffset);

        if (blockType != HapUtils::HAP_CODE_SIGN_BLOCK_ID) {
            PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR, "code sign data not exist in hap " + hapFilePath);
            return false;
        }
        auto ite = map.find(HapUtils::HAP_PROFILE_BLOCK_ID);
        if (ite == map.end())
            return false;
        ByteBuffer profileArray = ite->second;
        std::string profileArray_(profileArray.GetBufferPtr(), profileArray.GetCapacity());
        std::string profileContent;
        if (GetProfileContent(profileArray_, profileContent) < 0) {
            SIGNATURE_TOOLS_LOGE("get profile content failed, file: %s", hapFilePath.c_str());
            return false;
        }
        std::string suffix = fileNameArray[fileNameArray.size() - 1];
        bool isCodeSign = VerifyCodeSignature::VerifyHap(hapFilePath, blockOffset, blockLength,
                                                         suffix, profileContent);
        if (!isCodeSign) {
            SIGNATURE_TOOLS_LOGE("verify codesign failed, file: %s", hapFilePath.c_str());
            return false;
        }
        SIGNATURE_TOOLS_LOGI("verify codesign success.");
        return true;
    }
    SIGNATURE_TOOLS_LOGI("can not find codesign block.");
    return true;
}

int VerifyHap::GetProfileContent(const std::string profile, std::string& ret)
{
    json obj = json::parse(profile, nullptr, false);
    if (!obj.is_discarded() && obj.is_structured()) {
        ret = profile;
        return 0;
    }
    PKCS7Data p7Data;
    if (p7Data.Parse(profile) < 0) {
        ret = profile;
        return -1;
    }
    if (p7Data.Verify() < 0) {
        PrintErrorNumberMsg("PKCS7_VERIFY_ERROR", VERIFY_ERROR,
                            "Verify profile pkcs7 failed! Profile is invalid");
        ret = profile;
        return -1;
    }
    if (p7Data.GetContent(ret) < 0) {
        PrintErrorNumberMsg("PKCS7_VERIFY_ERROR", VERIFY_ERROR,
                            "Check profile failed, signed profile content is not byte array");
        ret = profile;
        return -1;
    }
    return 0;
}

bool VerifyHap::VerifyAppPkcs7(Pkcs7Context& pkcs7Context, const ByteBuffer& hapSignatureBlock)
{
    const unsigned char* pkcs7Block = reinterpret_cast<const unsigned char*>(hapSignatureBlock.GetBufferPtr());
    uint32_t pkcs7Len = static_cast<unsigned int>(hapSignatureBlock.GetCapacity());
    if (!VerifyHapOpensslUtils::ParsePkcs7Package(pkcs7Block, pkcs7Len, pkcs7Context)) {
        SIGNATURE_TOOLS_LOGE("parse pkcs7 failed");
        return false;
    }
    if (!VerifyHapOpensslUtils::GetCertChains(pkcs7Context.p7, pkcs7Context)) {
        SIGNATURE_TOOLS_LOGE("GetCertChains from pkcs7 failed");
        return false;
    }
    if (!VerifyHapOpensslUtils::VerifyPkcs7(pkcs7Context)) {
        SIGNATURE_TOOLS_LOGE("verify signature failed");
        return false;
    }
    return true;
}

bool VerifyHap::GetDigestAndAlgorithm(Pkcs7Context& digest)
{
    /*
     * contentinfo format:
     * int: version
     * int: block number
     * digest blocks:
     * each digest block format:
     * int: length of sizeof(digestblock) - 4
     * int: Algorithm ID
     * int: length of digest
     * byte[]: digest
     */
     /* length of sizeof(digestblock - 4) */
    int32_t digestBlockLen;
    if (!digest.content.GetInt32(DIGEST_BLOCK_LEN_OFFSET, digestBlockLen)) {
        SIGNATURE_TOOLS_LOGE("get digestBlockLen failed");
        return false;
    }
    /* Algorithm ID */
    if (!digest.content.GetInt32(DIGEST_ALGORITHM_OFFSET, digest.digestAlgorithm)) {
        SIGNATURE_TOOLS_LOGE("get digestAlgorithm failed");
        return false;
    }
    /* length of digest */
    int32_t digestlen;
    if (!digest.content.GetInt32(DIGEST_LEN_OFFSET, digestlen)) {
        SIGNATURE_TOOLS_LOGE("get digestlen failed");
        return false;
    }
    int32_t sum = sizeof(digestlen) + sizeof(digest.digestAlgorithm) + digestlen;
    if (sum != digestBlockLen) {
        SIGNATURE_TOOLS_LOGE("digestBlockLen: %d is not equal to sum: %d",
                             digestBlockLen, sum);
        return false;
    }
    /* set position to the digest start point */
    digest.content.SetPosition(DIGEST_OFFSET_IN_CONTENT);
    /* set limit to the digest end point */
    digest.content.SetLimit(DIGEST_OFFSET_IN_CONTENT + digestlen);
    digest.content.Slice();
    return true;
}

int32_t VerifyHap::WriteVerifyOutput(Pkcs7Context& pkcs7Context, std::vector<int8_t>& profile, Options* options)
{
    if (pkcs7Context.certChain.size() > 0) {
        bool flag = VerifyHap::HapOutPutCertChain(pkcs7Context.certChain[0],
            options->GetString(Options::OUT_CERT_CHAIN));
        if (!flag) {
            SIGNATURE_TOOLS_LOGE("out put cert chain failed");
            return IO_ERROR;
        }
    }
    if (pkcs7Context.p7 == nullptr) {
        std::string p7bContent(profile.begin(), profile.end());
        bool writeFlag = FileUtils::Write(p7bContent, options->GetString(Options::OUT_PROFILE)) < 0;
        if (writeFlag) {
            SIGNATURE_TOOLS_LOGE("p7b write to file falied!\n");
            return IO_ERROR;
        }
        return RET_OK;
    }
    bool pkcs7flag = VerifyHap::HapOutPutPkcs7(pkcs7Context.p7, options->GetString(Options::OUT_PROFILE));
    if (!pkcs7flag) {
        SIGNATURE_TOOLS_LOGE("out put p7b failed");
        return IO_ERROR;
    }
    return RET_OK;
}
} // namespace SignatureTools
} // namespace OHOS

File: OpenHarmony/developtools/hapsigner/hapsigntool/hap_sign_tool_lib/src/main/java/com/ohos/hapsigntool/hap/verify/VerifyHap.java
/*
 * Copyright (c) 2021-2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ohos.hapsigntool.hap.verify;

import com.ohos.hapsigntool.entity.Options;
import com.ohos.hapsigntool.codesigning.exception.FsVerityDigestException;
import com.ohos.hapsigntool.codesigning.exception.VerifyCodeSignException;
import com.ohos.hapsigntool.codesigning.sign.VerifyCodeSignature;
import com.ohos.hapsigntool.entity.Pair;
import com.ohos.hapsigntool.error.SignToolErrMsg;
import com.ohos.hapsigntool.hap.entity.SigningBlock;
import com.ohos.hapsigntool.error.HapFormatException;
import com.ohos.hapsigntool.error.ProfileException;
import com.ohos.hapsigntool.error.SignatureNotFoundException;
import com.ohos.hapsigntool.utils.FileUtils;
import com.ohos.hapsigntool.hap.utils.HapUtils;
import com.ohos.hapsigntool.entity.ParamConstants;
import com.ohos.hapsigntool.utils.LogUtils;
import com.ohos.hapsigntool.utils.StringUtils;
import com.ohos.hapsigntool.zip.ByteBufferZipDataInput;
import com.ohos.hapsigntool.zip.RandomAccessFileZipDataInput;
import com.ohos.hapsigntool.zip.UnsignedDecimalUtil;
import com.ohos.hapsigntool.zip.ZipDataInput;
import com.ohos.hapsigntool.zip.ZipFileInfo;
import com.ohos.hapsigntool.zip.ZipUtils;

import org.bouncycastle.cms.CMSException;
import org.bouncycastle.cms.CMSSignedData;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.bouncycastle.openssl.jcajce.JcaPEMWriter;
import org.bouncycastle.util.Arrays;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.OutputStream;
import java.io.RandomAccessFile;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.Security;
import java.security.cert.X509Certificate;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Class of verify hap.
 *
 * @since 2021/12/23
 */
public class VerifyHap {
    private static final LogUtils LOGGER = new LogUtils(VerifyHap.class);
    private static final int ZIP_HEAD_OF_SIGNING_BLOCK_LENGTH = 32;
    private static final int ZIP_HEAD_OF_SIGNING_BLOCK_COUNT_OFFSET_REVERSE = 28;
    private static final int ZIP_HEAD_OF_SUBSIGNING_BLOCK_LENGTH = 12;

    static {
        Security.addProvider(new BouncyCastleProvider());
    }

    private final boolean isPrintCert;

    public VerifyHap() {
        this(true);
    }

    public VerifyHap(boolean isPrintCert) {
        this.isPrintCert = isPrintCert;
    }

    private static String getProfileContent(byte[] profile) throws ProfileException {
        try {
            CMSSignedData cmsSignedData = new CMSSignedData(profile);
            if (!VerifyUtils.verifyCmsSignedData(cmsSignedData)) {
                throw new ProfileException(SignToolErrMsg.VERIFY_PROFILE_INVALID.toString());
            }
            Object contentObj = cmsSignedData.getSignedContent().getContent();
            if (!(contentObj instanceof byte[])) {
                throw new ProfileException(SignToolErrMsg.VERIFY_PROFILE_FAILED
                        .toString("Check profile failed, signed profile content is not byte array!"));
            }
            return new String((byte[]) contentObj, StandardCharsets.UTF_8);
        } catch (CMSException e) {
            return new String(profile, StandardCharsets.UTF_8);
        }
    }


    /**
     * Check whether parameters are valid
     *
     * @param options input parameters used to verify hap.
     * @return true, if all parameters are valid.
     */
    public boolean checkParams(Options options) {
        if (!options.containsKey(ParamConstants.PARAM_VERIFY_CERTCHAIN_FILE)) {
            LOGGER.error("Missing parameter: {}", ParamConstants.PARAM_VERIFY_CERTCHAIN_FILE);
            return false;
        }
        if (!options.containsKey(ParamConstants.PARAM_VERIFY_PROFILE_FILE)) {
            LOGGER.error("Missing parameter: {}", ParamConstants.PARAM_VERIFY_PROFILE_FILE);
            return false;
        }
        if (!options.containsKey(ParamConstants.PARAM_VERIFY_PROOF_FILE)) {
            LOGGER.warn("Missing parameter: {}", ParamConstants.PARAM_VERIFY_PROOF_FILE);
        }
        return true;
    }

    /**
     * verify hap file.
     *
     * @param options input parameters used to verify hap.
     * @return true, if verify successfully.
     */
    public boolean verify(Options options) {
        VerifyResult verifyResult;
        try {
            if (!checkParams(options)) {
                LOGGER.error("Check params failed!");
                throw new IOException();
            }
            String filePath = options.getString(ParamConstants.PARAM_BASIC_INPUT_FILE);
            if (StringUtils.isEmpty(filePath)) {
                LOGGER.error("Not found verify file path!");
                throw new IOException();
            }
            File signedFile = new File(filePath);
            if (!checkSignFile(signedFile)) {
                LOGGER.error("Check input signature hap false!");
                throw new IOException();
            }
            verifyResult = verifyHap(filePath);
            if (!verifyResult.isVerified()) {
                LOGGER.error("verify: {}", verifyResult.getMessage());
                throw new IOException();
            }
            String outputCertPath = options.getString(ParamConstants.PARAM_VERIFY_CERTCHAIN_FILE);
            if (verifyResult.getCertificates() != null) {
                writeCertificate(outputCertPath, verifyResult.getCertificates());
            }
        } catch (IOException e) {
            LOGGER.error("Write certificate chain error", e);
            return false;
        }

        String outputProfileFile = options.getString(ParamConstants.PARAM_VERIFY_PROFILE_FILE);
        String outputProofFile = options.getString(ParamConstants.PARAM_VERIFY_PROOF_FILE);
        String outputPropertyFile = options.getString(ParamConstants.PARAM_VERIFY_PROPERTY_FILE);
        try {
            outputOptionalBlocks(outputProfileFile, outputProofFile, outputPropertyFile, verifyResult);
        } catch (IOException e) {
            LOGGER.error("Output optional blocks error", e);
            return false;
        }

        LOGGER.info("verify: {}", verifyResult.getMessage());
        return true;
    }

    private void writeCertificate(String destFile, List<X509Certificate> certificates) throws IOException {
        try (JcaPEMWriter writer = new JcaPEMWriter(new FileWriter(destFile))) {
            for (final X509Certificate cert : certificates) {
                writer.write(cert.getSubjectDN().toString() + System.lineSeparator());
                writer.writeObject(cert);
            }
            LOGGER.info("Write certificate chain success!");
        }
    }

    private void outputOptionalBlocks(String outputProfileFile, String outputProofFile, String outputPropertyFile,
                                      VerifyResult verifyResult) throws IOException {
        List<SigningBlock> optionalBlocks = verifyResult.getOptionalBlocks();
        if (optionalBlocks != null && optionalBlocks.size() > 0) {
            for (SigningBlock optionalBlock : optionalBlocks) {
                int type = optionalBlock.getType();
                switch (type) {
                    case HapUtils.HAP_PROFILE_BLOCK_ID:
                        writeOptionalBytesToFile(optionalBlock.getValue(), outputProfileFile);
                        break;
                    case HapUtils.HAP_PROOF_OF_ROTATION_BLOCK_ID:
                        writeOptionalBytesToFile(optionalBlock.getValue(), outputProofFile);
                        break;
                    case HapUtils.HAP_PROPERTY_BLOCK_ID:
                        writeOptionalBytesToFile(optionalBlock.getValue(), outputPropertyFile);
                        break;
                    default:
                        throw new IOException("Unsupported Block Id: 0x" + Long.toHexString(type));
                }
            }
        }
        byte[] profile = verifyResult.getProfile();
        if (profile != null) {
            writeOptionalBytesToFile(profile, outputProfileFile);
        }
    }

    private void writeOptionalBytesToFile(byte[] data, String outputFile) throws IOException {
        if (outputFile == null || outputFile.isEmpty()) {
            return;
        }
        try (OutputStream out = Files.newOutputStream(Paths.get(outputFile))) {
            out.write(data);
            out.flush();
        }
    }

    private boolean checkSignFile(File signedFile) {
        try {
            FileUtils.isValidFile(signedFile);
        } catch (IOException e) {
            LOGGER.error("signedFile is invalid.", e);
            return false;
        }
        return true;
    }

    /**
     * Verify hap file.
     *
     * @param hapFilePath path of hap file.
     * @return true, if verify successfully.
     */
    public VerifyResult verifyHap(String hapFilePath) {
        VerifyResult result;
        try (RandomAccessFile fle = new RandomAccessFile(hapFilePath, "r")) {
            ZipDataInput hapFile = new RandomAccessFileZipDataInput(fle);
            ZipFileInfo zipInfo = ZipUtils.findZipInfo(hapFile);
            long eocdOffset = zipInfo.getEocdOffset();
            if (ZipUtils.checkZip64EoCDLocatorIsPresent(hapFile, eocdOffset)) {
                String errorMsg = "ZIP64 format not supported!";
                LOGGER.error(errorMsg);
                return new VerifyResult(false, VerifyResult.RET_UNSUPPORTED_FORMAT_ERROR, errorMsg);
            }
            HapUtils.HapSignBlockInfo hapSigningBlockAndOffsetInFile = HapUtils.findHapSigningBlock(hapFile, zipInfo);
            ByteBuffer signingBlock = hapSigningBlockAndOffsetInFile.getContent();
            signingBlock.order(ByteOrder.LITTLE_ENDIAN);
            Pair<ByteBuffer, List<SigningBlock>> blockPair = getHapSignatureSchemeBlockAndOptionalBlocks(signingBlock);
            ByteBuffer signatureSchemeBlock = blockPair.getFirst();
            List<SigningBlock> optionalBlocks = blockPair.getSecond();
            Collections.reverse(optionalBlocks);
            if (!checkCodeSign(hapFilePath, optionalBlocks)) {
                String errMsg = "code sign verify failed";
                return new VerifyResult(false, VerifyResult.RET_CODESIGN_DATA_ERROR, errMsg);
            }
            HapVerify verifyEngine = getHapVerify(hapFile, zipInfo, hapSigningBlockAndOffsetInFile,
                    signatureSchemeBlock, optionalBlocks);
            result = verifyEngine.verify();
            result.setSignBlockVersion(hapSigningBlockAndOffsetInFile.getVersion());
        } catch (IOException e) {
            LOGGER.error("Verify Hap has IO error!", e);
            result = new VerifyResult(false, VerifyResult.RET_IO_ERROR, e.getMessage());
        } catch (SignatureNotFoundException e) {
            LOGGER.error("Verify Hap failed, signature not found.", e);
            result = new VerifyResult(false, VerifyResult.RET_SIGNATURE_NOT_FOUND_ERROR, e.getMessage());
        } catch (HapFormatException e) {
            LOGGER.error("Verify Hap failed, unsupported format hap.", e);
            result = new VerifyResult(false, VerifyResult.RET_UNSUPPORTED_FORMAT_ERROR, e.getMessage());
        } catch (FsVerityDigestException e) {
            LOGGER.error("Verify Hap failed, fs-verity digest generate failed.", e);
            result = new VerifyResult(false, VerifyResult.RET_DIGEST_ERROR, e.getMessage());
        } catch (VerifyCodeSignException e) {
            LOGGER.error("Verify Hap failed, code sign block verify failed.", e);
            result = new VerifyResult(false, VerifyResult.RET_CODE_SIGN_BLOCK_ERROR, e.getMessage());
        } catch (CMSException e) {
            LOGGER.error("Verify Hap failed, code signature verify failed.", e);
            result = new VerifyResult(false, VerifyResult.RET_SIGNATURE_ERROR, e.getMessage());
        } catch (ProfileException e) {
            LOGGER.error("Verify Hap failed, parse app-identifier from profile failed, profile is invalid", e);
            return new VerifyResult(false, VerifyResult.RET_CODE_SIGN_BLOCK_ERROR, e.getMessage());
        }
        return result;
    }

    private HapVerify getHapVerify(ZipDataInput hapFile, ZipFileInfo zipInfo,
                                   HapUtils.HapSignBlockInfo hapSigningBlockAndOffsetInFile,
                                   ByteBuffer signatureSchemeBlock, List<SigningBlock> optionalBlocks) {
        long signingBlockOffset = hapSigningBlockAndOffsetInFile.getOffset();
        ZipDataInput beforeHapSigningBlock = hapFile.slice(0, signingBlockOffset);
        ZipDataInput centralDirectoryBlock = hapFile.slice(zipInfo.getCentralDirectoryOffset(),
                zipInfo.getCentralDirectorySize());
        ByteBuffer eocdBbyteBuffer = zipInfo.getEocd();
        ZipUtils.setCentralDirectoryOffset(eocdBbyteBuffer, signingBlockOffset);
        ZipDataInput eocdBlock = new ByteBufferZipDataInput(eocdBbyteBuffer);
        HapVerify verifyEngine = new HapVerify(beforeHapSigningBlock, signatureSchemeBlock,
                centralDirectoryBlock, eocdBlock, optionalBlocks);
        verifyEngine.setIsPrintCert(isPrintCert);
        return verifyEngine;
    }

    /**
     * code sign check
     *
     * @param hapFilePath hap file path
     * @param optionalBlocks optional blocks
     * @return true or false
     * @throws FsVerityDigestException FsVerity digest on error
     * @throws IOException IO error
     * @throws VerifyCodeSignException verify code sign on error
     * @throws CMSException cms on error
     * @throws ProfileException profile of the hap error
     */
    private boolean checkCodeSign(String hapFilePath, List<SigningBlock> optionalBlocks)
            throws FsVerityDigestException, IOException, VerifyCodeSignException, CMSException, ProfileException {
        Map<Integer, byte[]> map = optionalBlocks.stream()
                .collect(Collectors.toMap(SigningBlock::getType, SigningBlock::getValue));
        byte[] propertyBlockArray = map.get(HapUtils.HAP_PROPERTY_BLOCK_ID);
        if (propertyBlockArray != null && propertyBlockArray.length > 0) {
            LOGGER.info("trying verify codesign block");
            String[] fileNameArray = hapFilePath.split("\\.");
            if (fileNameArray.length < ParamConstants.FILE_NAME_MIN_LENGTH) {
                LOGGER.error("ZIP64 format not supported");
                return false;
            }
            ByteBuffer byteBuffer = ByteBuffer.wrap(propertyBlockArray);
            ByteBuffer header = HapUtils.reverseSliceBuffer(byteBuffer, 0, ZIP_HEAD_OF_SUBSIGNING_BLOCK_LENGTH);
            long blockOffset = UnsignedDecimalUtil.getUnsignedInt(header);
            int blockLength = header.getInt();
            int blockType = header.getInt();
            if (blockType != HapUtils.HAP_CODE_SIGN_BLOCK_ID) {
                LOGGER.error("Verify Hap has no code sign data error!");
                return false;
            }
            File outputFile = new File(hapFilePath);
            byte[] profileArray = map.get(HapUtils.HAP_PROFILE_BLOCK_ID);
            String profileContent = getProfileContent(profileArray);
            String suffix = fileNameArray[fileNameArray.length - 1];
            boolean isCodeSign = VerifyCodeSignature.verifyHap(outputFile, blockOffset, blockLength,
                    suffix, profileContent);
            if (!isCodeSign) {
                LOGGER.error("Verify Hap has no code sign data error!");
                return false;
            }
            LOGGER.info("verify codesign success");
            return true;
        }
        LOGGER.info("can not find codesign block");
        return true;
    }

    private Pair<ByteBuffer, List<SigningBlock>> getHapSignatureSchemeBlockAndOptionalBlocks(ByteBuffer hapSigningBlock)
            throws SignatureNotFoundException {
        try {
            ByteBuffer header = HapUtils.reverseSliceBuffer(
                    hapSigningBlock,
                    hapSigningBlock.capacity() - ZIP_HEAD_OF_SIGNING_BLOCK_LENGTH,
                    hapSigningBlock.capacity());
            ByteBuffer value = HapUtils.reverseSliceBuffer(hapSigningBlock, 0,
                    hapSigningBlock.capacity() - ZIP_HEAD_OF_SIGNING_BLOCK_LENGTH);

            byte[] signatureValueBytes = new byte[value.capacity()];
            value.get(signatureValueBytes, 0, signatureValueBytes.length);
            signatureValueBytes = Arrays.reverse(signatureValueBytes);
            header.position(ZIP_HEAD_OF_SIGNING_BLOCK_COUNT_OFFSET_REVERSE); // position to the block count offset
            int blockCount = header.getInt();
            int current = value.capacity() - ZIP_HEAD_OF_SUBSIGNING_BLOCK_LENGTH * blockCount;
            value.position(current);

            int blockType = -1;
            int blockLength = -1;
            int blockOffset = -1;
            ByteBuffer hapSigningPkcs7Block = null;
            List<SigningBlock> optionalBlocks = new ArrayList<SigningBlock>();
            for (int i = 0; i < blockCount; i++) {
                blockOffset = value.getInt();
                blockLength = value.getInt();
                blockType = value.getInt();
                if (blockOffset + blockLength > signatureValueBytes.length) {
                    throw new SignatureNotFoundException("block end pos: " + (blockOffset + blockLength)
                            + " is larger than block len: " + signatureValueBytes.length);
                }
                if (HapUtils.getHapSignatureOptionalBlockIds().contains(blockType)) {
                    byte[] blockValue = Arrays.copyOfRange(signatureValueBytes, blockOffset, blockOffset + blockLength);
                    optionalBlocks.add(new SigningBlock(blockType, blockValue));
                }
                if (blockType == HapUtils.HAP_SIGNATURE_SCHEME_V1_BLOCK_ID) {
                    byte[] result = Arrays.copyOfRange(signatureValueBytes, blockOffset, blockOffset + blockLength);
                    hapSigningPkcs7Block = ByteBuffer.wrap(result);
                }
            }
            return Pair.create(hapSigningPkcs7Block, optionalBlocks);
        } finally {
            hapSigningBlock.clear();
        }
    }
}